sockeye/loss.py

```diff
@@ -145,40 +145,48 @@ def __init__(self,
         self.loss_config = loss_config
 
     @staticmethod
-    def cross_entropy(pred, label, ignore):
-        prob = mx.nd.pick(pred, label.astype(dtype="int32"))
-        prob = prob * (1 - ignore) + ignore
-        loss = -mx.nd.log(prob + 1e-8)  # pylint: disable=invalid-unary-operand-type
-        return loss
-
-    def cross_entropy_smoothed(self, pred, label, ignore):
-        label_dist = mx.nd.one_hot(indices=label.astype(dtype='int32'),
-                                   depth=self.loss_config.vocab_size,
-                                   on_value=1.0 - self.loss_config.label_smoothing,
-                                   off_value=self.loss_config.label_smoothing / (self.loss_config.vocab_size - 1.0))
-        label_dist = mx.nd.where(1 - ignore, label_dist, mx.nd.zeros_like(label_dist))
-        loss = label_dist * (- mx.nd.log(pred + 1e-8))  # pylint: disable=invalid-unary-operand-type
-        return loss
+    def cross_entropy(logprob, label):
+        ce = -mx.nd.pick(logprob, label)  # pylint: disable=invalid-unary-operand-type
+        return ce
+
+    @staticmethod
+    def cross_entropy_smoothed(logprob, label, alpha, num_classes):
+        ce = CrossEntropyMetric.cross_entropy(logprob, label)
+        # gain for each incorrect class
+        per_class_gain = alpha / (num_classes - 1)
+        # discounted loss for correct class
+        ce *= 1 - alpha - per_class_gain
+        # add gain for incorrect classes to total cross-entropy
+        ce -= mx.nd.sum(logprob * per_class_gain, axis=-1, keepdims=False)
+        return ce
 
     def update(self, labels, preds):
         for label, pred in zip(labels, preds):
             batch_size = label.shape[0]
             label = label.as_in_context(pred.context).reshape((label.size,))
-            # Ignore padding
-            # TODO: contribute ignoring padding for cross-entropy back to MXNet
-            ignore = (label == C.PAD_ID).astype(dtype=pred.dtype)
 
+            logprob = mx.nd.log(pred + 1e-8)
```

Comment:
would it make sense to do this with `mx.sym.clip`?

Reply:
Ok, will use the same pattern as in MXNET's perplexity metric: https://github.com/apache/incubator-mxnet/blob/master/python/mxnet/metric.py#L749